File: ollama_proxy.py
from http.server import HTTPServer, BaseHTTPRequestHandler
import urllib.request
import json
import logging

logging.basicConfig(level=logging.INFO, format='%(message)s')

class Proxy(BaseHTTPRequestHandler):
    def do_POST(self):
        content_length = int(self.headers['Content-Length'])
        post_data = self.rfile.read(content_length)
        
        req = urllib.request.Request("http://localhost:11434" + self.path, data=post_data, headers=dict(self.headers), method="POST")
        try:
            with urllib.request.urlopen(req) as response:
                body = response.read()
                
                # Check if it's the chat endpoint to log it
                if "/api/chat" in self.path or "/api/generate" in self.path:
                    logging.info("=== PROXY INTERCEPTED REQUEST TO OLLAMA ===")
                    try:
                        req_json = json.loads(post_data.decode('utf-8'))
                        logging.info("Model: %s", req_json.get("model"))
                        messages = req_json.get("messages", [])
                        logging.info("Messages Array: %s", json.dumps(messages, indent=2))
                        
                        total_chars = sum(len(m.get("content", "")) for m in messages)
                        logging.info("Total Characters in Messages: %d", total_chars)
                        
                    except Exception as e:
                        logging.info("Could not parse request JSON: %s", e)
                        
                    logging.info("=== OLLAMA RAW RESPONSE ===")
                    logging.info(body.decode('utf-8'))
                    logging.info("===========================================")
                
                self.send_response(response.status)
                for k, v in response.headers.items():
                    self.send_header(k, v)
                self.end_headers()
                self.wfile.write(body)
        except Exception as e:
            self.send_response(500)
            self.end_headers()
            self.wfile.write(str(e).encode('utf-8'))

if __name__ == '__main__':
    server = HTTPServer(('localhost', 11435), Proxy)
    print("Proxy started on port 11435")
    server.serve_forever()

File: ollama_proxy.js
const http = require('http');

const PORT = 11435;
const TARGET_HOST = 'localhost';
const TARGET_PORT = 11434;

const server = http.createServer((clientReq, clientRes) => {
    let body = '';
    
    // Determine if it's the chat endpoint
    const isChat = clientReq.url.includes('/api/chat') || clientReq.url.includes('/api/generate');

    if (isChat && clientReq.method === 'POST') {
        clientReq.on('data', chunk => {
            body += chunk.toString();
        });
        clientReq.on('end', () => {
            console.log("\n=== PROXY INTERCEPTED REQUEST TO OLLAMA ===");
            try {
                const reqJson = JSON.parse(body);
                console.log("Model: " + reqJson.model);
                console.log("Messages Array: " + JSON.stringify(reqJson.messages, null, 2));
                let totalChars = 0;
                if (reqJson.messages) {
                    reqJson.messages.forEach(m => {
                        totalChars += (m.content || "").length;
                    });
                }
                console.log("Total Characters in Messages: " + totalChars);
            } catch (e) {
                console.log("Could not parse request JSON: " + e.message);
            }
            
            // Forward the collected body to Ollama
            const options = {
                hostname: TARGET_HOST,
                port: TARGET_PORT,
                path: clientReq.url,
                method: clientReq.method,
                headers: clientReq.headers
            };
            
            // Adjust headers for our collected body
            options.headers['content-length'] = Buffer.byteLength(body);
            delete options.headers['transfer-encoding'];

            const proxyReq = http.request(options, proxyRes => {
                let responseBody = '';
                
                // Write headers to client
                clientRes.writeHead(proxyRes.statusCode, proxyRes.headers);
                
                proxyRes.on('data', chunk => {
                    responseBody += chunk.toString();
                    clientRes.write(chunk); // stream to client
                });
                
                proxyRes.on('end', () => {
                    console.log("\n=== OLLAMA RAW RESPONSE ===");
                    console.log(responseBody);
                    console.log("===========================================\n");
                    clientRes.end();
                });
            });
            
            proxyReq.on('error', err => {
                console.error("Proxy request error: ", err);
                clientRes.writeHead(500);
                clientRes.end();
            });
            
            proxyReq.write(body);
            proxyReq.end();
        });
    } else {
        // Just forward everything else directly without body parsing
        const options = {
            hostname: TARGET_HOST,
            port: TARGET_PORT,
            path: clientReq.url,
            method: clientReq.method,
            headers: clientReq.headers
        };
        const proxyReq = http.request(options, proxyRes => {
            clientRes.writeHead(proxyRes.statusCode, proxyRes.headers);
            proxyRes.pipe(clientRes, { end: true });
        });
        clientReq.pipe(proxyReq, { end: true });
        
        proxyReq.on('error', err => {
            console.error("Proxy direct request error: ", err);
            clientRes.writeHead(500);
            clientRes.end();
        });
    }
});

server.listen(PORT, () => {
    console.log(`Proxy listening on port ${PORT}`);
});
